apptest/app.go

```diff
@@ -177,14 +182,42 @@ func (app *app) processOutput(outputName string, output io.Reader, lps ...linePr
 	}
 }
 
-// writeToStderr is a line processor that writes the line to the stderr.
+// captureStdout is a lineProcessor that writes the line to the stderr.
 // The function always returns false to indicate its caller that each line must
 // be written to the stderr.
-func (app *app) writeToStderr(line string) bool {
-	fmt.Fprintf(os.Stderr, "%s %s\n", app.instance, line)
+func (app *app) captureStdout(line string) bool {
+	app.outputLock.Lock()
+	defer app.outputLock.Unlock()
+	app.stdout = append(app.stdout, line)
 	return false
 }
 
+// captureStderr is a lineProcessor that writes the line to the stderr.
+// The function always returns false to indicate its caller that each line must
+// be written to the stderr.
+func (app *app) captureStderr(line string) bool {
+	app.outputLock.Lock()
+	defer app.outputLock.Unlock()
+	app.stderr = append(app.stderr, line)
+	return false
+}
+
+func (app *app) FlushOutput() {
+	app.outputLock.Lock()
+	defer app.outputLock.Unlock()
+
+	for _, line := range app.stdout {
+		fmt.Fprintf(os.Stderr, "%s %s\n", app.instance, line)
+	}
+
+	for _, line := range app.stderr {
```

Comment:
this could break the original order. I think we should store stdout and stderr into one []stirng slice